radicle-httpd/src/api.rs

```diff
@@ -115,6 +115,14 @@ pub struct PaginationQuery {
     pub per_page: Option<usize>,
 }
 
+#[derive(Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct CobsQuery {
+    pub page: Option<usize>,
+    pub per_page: Option<usize>,
+    pub state: Option<String>,
```

Comment:
I'm wondering if this should be:
```suggestion
    pub state: Option<State>,
```

You wouldn't need the `From` or `FromStr` and leave it up to `Deserialize` to handle the conversion.

Reply:
Yeah thats what I also wanted to do in the beginning, but since we serialize `State` to `{ "status": "open" }` I would need to pass that through a querystring which is not what I was after.

My goal was to pass e.g. `?state=open` and for the endpoint to parse it into `patch::State::Open`
Also wrote a custom deserializer, but in the end I thought it to be easier to convert the passed `String` into `State`

Reply:
Wait, how does the deserializer work? :thinking: does it not accept `?state=open`? I'm not sure I'm understanding what the problem is

Reply:
as far as I understand, the serializer serializes `patch::State::Open` into `{ "status": "open" }` due to the serde container tag attribute on `State` `#[serde(rename_all = "camelCase", tag = "status")]`
for reference: https://serde.rs/container-attrs.html#tag
So the deserializer expects a `{ "status": "open" }` to be able to parse it back to `patch::State::Open`
Or maybe I'm missing something with serde.

Reply:
Ahhh I see... Ya that's annoying. I tried googling for any solutions to this, but I can't seem to see any. It might be worth asking in the `axum` project if there's a way of doing this

Reply:
Ok I think this commit does the trick
6ec2614

Reply:
I've asked them here https://github.com/tokio-rs/axum/discussions/1945

Reply:
Ok, I think we can do the following:
* Implement `FromStr` for the `State` operations
* Change this field to `pub state: Option<State>`
* Add the pragma `#[serde(deserialize_with="from_str")]`

Could you see if that works? I'm unsure if the `Option` will be a problem with the pragma or not

Reply:
Ok I tried that and it seems that `Option` is a problem with the pragma.
And IMO by doing that we also change the `Deserialize` of `State` which isn't inverse to `Serialize` anymore.

So either we accept strings and parse them, or write our own Deserializer.

Reply:
Or we could also remove the `tag = "status"` which I don't think makes much sense now..
If at some point we want to show more information under the `state` property we could certainly revisit this, since we use `serde` mostly for serialization for `radicle-httpd` right?

Reply:
>  And IMO by doing that we also change the Deserialize of State which isn't inverse to Serialize anymore.

It wouldn't change the `Deserialize` of `State`, it would be specifying that `QueryCobs` is using a different way of deserializing `State` :)

 > So either we accept strings and parse them, or write our own Deserializer.
 
 Thanks for trying, I'd say we go with accept strings and use `FromStr`. I was hoping this version would work. A very last alternative is to define a standalone function:
 ```rust
 fn parse_state(state: String) -> Result<Option<State>, String> {
     if state.is_empty() { None } else { state.parse().map(Some).map_err(|err| err.to_string())
 }
 ```
and use `#[serde(deserialize_with = "parse_state")]`

> Or we could also remove the tag = "status" which I don't think makes much sense now..

Aye, why did you opt for this @cloudhead? Removing it now will break backwards compatibility though and we're now in production ;)

Reply:
> > So either we accept strings and parse them, or write our own Deserializer.
> 
> Thanks for trying, I'd say we go with accept strings and use `FromStr`. I was hoping this version would work. A very last alternative is to define a standalone function:
> 
> ```rust
> fn parse_state(state: String) -> Result<Option<State>, String> {
>     if state.is_empty() { None } else { state.parse().map(Some).map_err(|err| err.to_string())
> }
> ```
> 
> and use `#[serde(deserialize_with = "parse_state")]`

Actually, this is quite nice. It's basically equivalent to the parsing solution but we localise the parsing at the point of deserialization instead of in the handler :)